ocrd/ocrd/workspace_bagger.py

```diff
@@ -80,7 +81,10 @@ def _bag_mets_files(
                     file_grp_dir.mkdir()
 
                 attr = 'local_filename' if f.local_filename else 'url'
-                basename = f.basename if f.basename else f"{f.ID}{MIME_TO_EXT.get(f.mimetype, '.xml')}"
+                if f.local_filename and f.basename:
+                    basename = f.basename
+                else:
+                    basename = safe_filename(f.contentids if f.contentids else f.ID) + MIME_TO_EXT.get(f.mimetype, '.xml')
```

Comment:
I would advise against direct use of `@CONTENTIDS` as file name. The URL prefix almost always is not what you want. How about stripping the host-name part (if in fact it is a URL), and then using `makedirs` for all remaining path prefixes?

Reply:
Fair points. Agree that representing the URL path as directory is a neat way to do it, though it deviates from our general flat directory structure below the `fileGrp` dirs. Removing the host is also prettier.

But I'm wondering how that would work for @M3ssman's use case - how much info do you need to still be able to debug your workflows?

Reply:
It is not intended to be used as-it-is, simply because an URN/URI contains chars that are not valid as part in local filenames. Therefore - at least in ULB and semantics workflows - colon is exchanged with a plus sign.

Consider a page container like this

```
<mets:div ID="phys1278993" TYPE="page" CONTENTIDS="urn:nbn:de:gbv:3:1-113129-p0007-8" ORDER="1">
        <mets:fptr FILEID="IMG_MAX_1278993"/>
        <mets:fptr FILEID="OCR-D-GT-FULLTEXT-1"/>
</mets:div>

```
with GT linked

```
<mets:file ID="OCR-D-GT-FULLTEXT-1" MIMETYPE="application/vnd.prima.page+xml">
	<mets:FLocat xlink:href="GT-PAGE/urn+nbn+de+gbv+3+1-113129-p0007-8_ger.gt.xml" LOCTYPE="OTHER" OTHERLOCTYPE="FILE"/>
</mets:file>
```

but the actual imsge via URL like this

```
<mets:file MIMETYPE="image/jpeg" ID="IMG_MAX_1278993">
	<mets:FLocat xlink:href="https://opendata.uni-halle.de/retrieve/eeeee05d-c7cd-4e89-9607-5d1ac175afa1/00000007.jpg" LOCTYPE="URL"/>
</mets:file>
```

The goal is to match both files, GT and Image, by their names alone without any extensions. 
If this can be achieved, both can be used out-of-the-box for further GT-works in Tools like Transkribus or Larex.
(I have to handle 1.600 GT-ODEM-files, nearly 100 newspaper-GT-files, 101 GT arabic and about 400 pages GT persian. And if our next digi-project will be granted, the GT will increase further.)

Proposal: Instead of using the `CONTENTIDS` attribute it'd be sufficient to rename the images locally like the corresponding GT-file, whatever it's name was. My main concern is therefore to have equal names, not to name something like some attribute.

Maybe this way one hopefully avoids additional processing?





(c.f. https://github.com/M3ssman/gt-test/blob/a370f3a691506f4eab1b91226a76a7c1f461ba10/data/corpus-odem-ger-256/mets.xml#L8905)

Reply:
> It is not intended to be used as-it-is, simply because an URN/URI contains chars that are not valid as part in local filenames. Therefore - at least in ULB and semantics workflows - colon is exchanged with a plus sign.

It is not used as-is but passed through `safe_filename` which does

```python
def safe_filename(url):                                              
    """                                                              
    Sanitize input to be safely used as the basename of a local file.
    """                                                              
    ret = re.sub(r'[^\w]+', '_', url)                                
    ret = re.sub(r'^\.*', '', ret)                                   
    ret = re.sub(r'\.\.*', '.', ret)                                 
    #  print('safe filename: %s -> %s' % (url, ret))                 
    return ret                                                       
```

Adapting this to make the replacement (`_`) configurable as `+` is not an issue. However

> Proposal: Instead of using the `CONTENTIDS` attribute it'd be sufficient to rename the images locally like the corresponding GT-file, whatever it's name was. My main concern is therefore to have equal names, not to name something like some attribute.

I still don't understand how to achieve that. For your example

```xml
<mets:file ID="OCR-D-GT-FULLTEXT-1" MIMETYPE="application/vnd.prima.page+xml">
	<mets:FLocat xlink:href="GT-PAGE/urn+nbn+de+gbv+3+1-113129-p0007-8_ger.gt.xml" LOCTYPE="OTHER" OTHERLOCTYPE="FILE"/>
</mets:file>
...
<mets:file MIMETYPE="image/jpeg" ID="IMG_MAX_1278993">
	<mets:FLocat xlink:href="https://opendata.uni-halle.de/retrieve/eeeee05d-c7cd-4e89-9607-5d1ac175afa1/00000007.jpg" LOCTYPE="URL"/>
</mets:file>
...
<mets:div ID="phys1278993" TYPE="page" CONTENTIDS="urn:nbn:de:gbv:3:1-113129-p0007-8" ORDER="1">
        <mets:fptr FILEID="IMG_MAX_1278993"/>
        <mets:fptr FILEID="OCR-D-GT-FULLTEXT-1"/>
</mets:div>
```

What should the bagger write as the filenames of `IMG_MAX_1278993` and `OCR-D-GT-FULLTEXT-1`?

Reply:
I'd wish to save `urn+nbn+de+gbv+3+1-113129-p0007-8_ger.gt.xml` as filename for the GT, since this is the file which `OCR-D-GT-FULLTEXT-1` points to, and for the image where container `IMG_MAX_1278993` refers to, a corresponding name like `urn+nbn+de+gbv+3+1-113129-p0007-8_ger.jpg` , if it's possible.

Reply:
The point at least in my understanding - ALIMU- concerning ULB is, that the GT-files shall be published (and probably edited further afterwards) but will be tied to the GT-repository. There is nothing about to change with the images, they only need to be referenced and resolvable, for example, for later generation of training data using the GT-files.